### DISCUSSION TITLE
Any suggestions on how to extract contextual named entities by rule matcher?

### DISCUSSION THREAD

**Discussion by lingvisa** [explosion/spaCy#6610] [Help: Coding & Implementations]
For exaple:

`Xyz Abc is a great product!.`

My rule may look like this:

` [{"POS": {"IN": ["NN", "NNP"]}}, {"POS": {"IN": ["NN", "NNP"]}, "OP": "*"},  { "OP": "+"},   { "TEXT": {"IN": ["product"]}}]`

In this case the string will be extracted as a Product below:

`[Xyz Abc is a great product]`

However, 'is a great product' is just the context, which should be excluded from the string. Is there a way to just extract Xyz Abc as a product? If I look for the first context word 'is' and chop off from there from the matched token list, this will work for this case, but there are many other cases in which the contexts take different forms. If there a way to make the start and end of the exact NE string in the pattern, that would be great, since many NE token rules need context checking like this. 

Any suggestions on how to tackle this problem?

**Comment by lingvisa** [comment]
Another nice thing would be extract multiple entities from one pattern. For example:

` Xyz Abc is a great cosmetic product!`

Not only 'Xyz Abc' is a Product, but also it is a 'Cosmetic' product. I want to extract both in one pattern. Otherwise, almost the same pattern needs to be repeated for two types of entities.

**Comment by svlandeg** [comment]
Hi! Just FYI - I converted your issue to a discussion thread. We're aiming to keep the Issue tracker focused mainly on bug reports, while the new GH Discussion board can be a place where the community can come together, help eachother, discuss best practices, and so on.

**Comment by lingvisa** [comment]
`Microsoft is a good company`

`[{"POS": {"IN": ["NNP"]}}, { "LEMMA": "be"}], { "OP": "+"}, { "TEXT": "company"}]`

To exactly extract 'Microsoft' as the string of NE, how likely to introduce a pre/post context marker in the predicates of the pattern? For the above case, it can be enhanced like below:

 `[{"POS": {"IN": ["NNP"]}}, { "LEMMA": "be", "POST_CONTEXT": true}, { "OP": "+"}, { "TEXT": "company"}]`

The 'POST_CONTEXT' key indicates that the token 'be' will be the **first** token of the post context "is a good company", and should be excluded from the NE string.

Likewise, for this sentence:
The software company Microsoft launched a new cloud project.
`[{"LEMMA": "the"}, { "OP": "+"}, { "TEXT": "company", "PRE_CONTEXT": true}, {"POS": {"IN": ["NNP"]}}]`

The 'PRE_CONTEXT' key indicates that the token 'company' will be the **last** token of the pre context "The software company", and should be excluded from the NE string.

How hard to introduce these two keys or something like that in the current rule matcher? Context checking is essential for serious pattern matchings in NE extraction. Why not do this in python code with parse matching?  First, in many cases to match on parse tree is not as robust as on linear string, especially for non-English languages. Second, even for English, to do pattern matching in code doesn't scale well, especially if your information extraction tasks are diverse. And the 3rd avenue is through models and that's true. But to make the rule matcher more useful, to be able to mark some type of context checking as shown above is really important.
